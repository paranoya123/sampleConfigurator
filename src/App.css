

.App label{
  display: block;
  margin: 10px;
}

.App label select, .App label input{
  margin-left: 10px;
}