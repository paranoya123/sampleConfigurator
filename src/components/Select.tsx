import React from "react";

export function Select(props: {name: string,
    options: Array<{name: string}>,
    selected: number|string,
    onChange: (e: React.ChangeEvent<HTMLSelectElement>) => void})
{
    const {name, options, selected, onChange} = props
    return (
        <label>
            {name}
            <select value={selected} onChange={onChange}>
                {options.map((o, i)=>(
                    <option key = {i} value = {i}> {o.name} </option>
                ))}
            </select>
        </label>
    )
}