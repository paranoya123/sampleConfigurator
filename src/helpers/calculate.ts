export const calculatePrice =
    (colorId: number, finenessId: number, size: number, engravingLength: number, matrix: Array<Array<number>>): number =>
    (matrix[colorId][finenessId]) * size + engravingLength * 10