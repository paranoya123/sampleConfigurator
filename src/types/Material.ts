import Color from './Color'
import Fineness from './Fineness'


export default interface Material {
    name: string,
    colors: Array<Color>,
    fineness: Array<Fineness>
}