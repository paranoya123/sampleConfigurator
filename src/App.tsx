import React, {useState, useEffect} from 'react'
import './App.css'
import {Select} from './components/Select'
import Material from './types/Material'
import Fineness from './types/Fineness'
import Color from './types/Color'
import {calculatePrice} from './helpers/calculate'
import Prices from "./types/Prices"

function App(props: {materials: Array<Material>, sizes: Array<{name: string}>,  matrix: Array<Prices>}) {
  const {materials, sizes, matrix} = props
  const [selectedMaterial, setSelectedMaterial] = useState(0)
  const [selectedColor, setSelectedColor] = useState(0)
  const [selectedFineness, setSelectedFineness] = useState(0)
  const [engraving, setEngraving] = useState('')
  const [selectedSize, setSelectedSize] = useState(0)
  const [price, setPrice] = useState(0)


  const onChangeMaterial = (e: React.ChangeEvent<HTMLSelectElement>) => {
      const index = e.target.value
      setSelectedMaterial(+index)
      setSelectedColor(0)
      setSelectedFineness(0)
  }
  const onChangeColor = (e: React.ChangeEvent<HTMLSelectElement>) => {
      const index = e.target.value
      setSelectedColor(+index)
  }

  const onChangeFineness = (e: React.ChangeEvent<HTMLSelectElement>) => {
      const index = e.target.value
      setSelectedFineness(+index)
  }

  const onChangeEngraving = (e: React.ChangeEvent<HTMLInputElement>) => {
      const engraving = e.target.value
      setEngraving(engraving)
  }

  const onChangeSize = (e: React.ChangeEvent<HTMLSelectElement>) => {
      const index = e.target.value
      setSelectedSize(+index)
  }

  useEffect(() => {
      const engravingLength = engraving.replace(/\s/g, '').length
      const m = matrix.find(material => materials[selectedMaterial].name === material.material)
      if(!m) return
      const price = calculatePrice(materials[selectedMaterial].colors[selectedColor].name,
                        materials[selectedMaterial].fineness[selectedFineness].name,
                        +sizes[selectedSize].name,
                        engravingLength,
                        m
      )
      setPrice(price)
   }, [engraving, selectedSize, selectedFineness, selectedColor, selectedMaterial, sizes, matrix])

  return (
      <div className={'App'}>
        <Select
            name={'Material'}
            options={materials}
            selected = {selectedMaterial}
            onChange = {onChangeMaterial}
        />
        <Select
            name={'Color'}
            options={materials[selectedMaterial].colors}
            selected = {selectedColor}
            onChange = {onChangeColor}
        />
        <Select
            name={'Fineness'}
            options={materials[selectedMaterial].fineness}
            selected = {selectedFineness}
            onChange = {onChangeFineness}
        />
        <Select
            name={'Sizes'}
            options={sizes}
            selected = {selectedSize}
            onChange = {onChangeSize}
        />

        <label>
            Engraving :
            <input id={'engraving'} data-testid="input" type="text" value={engraving} onChange={onChangeEngraving} />
        </label>

        <label id={'price'} >Price: {price}</label>
      </div>
  )
}

export default App
