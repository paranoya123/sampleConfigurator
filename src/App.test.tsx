import React from 'react'
import App from './App'
import {Select} from './components/Select'
import Material from "./types/Material"
import {calculatePrice} from './helpers/calculate'
import renderer from 'react-test-renderer'
import { render, screen, fireEvent, getByText } from '@testing-library/react'

const priceMatrixGold = [[1000, 200, 300], [800, 900, 100], [500, 800, 1100]]
const priceMatrixPlatinum = [[1000, 2000, 3000], [8000, 9000, 1000], [5000, 8000, 11000]]
const priceMatrixSilver = [[100, 20, 30], [80, 90, 10], [50, 80, 110]]

const colorsForGold = [{name: 'Red'}, {name: 'Green'} ,{name: 'Blue'}]
const colorsForPlatinum = [{name: 'White'}, {name: 'Yellow'} ,{name: 'Blue'}]
const colorsForSilver = [{name: 'Black'}, {name: 'Green'} ,{name: 'Grey'}]

const finenessForGold = [{name: '500'}, {name: '585'} ,{name: '700'}]
const finenessForPlatinum = [{name: '322'}, {name: '888'} ,{name: '444'}]
const finenessForSilver = [{name: '323'}, {name: '898'} ,{name: '999'}]


const data: Array<Material> = [{name: 'Gold', colors: colorsForGold, fineness: finenessForGold},
  {name: 'Platinum', colors: colorsForPlatinum, fineness: finenessForPlatinum},
  {name: 'Silver', colors: colorsForSilver, fineness: finenessForSilver}]

const sizes  = [{name: '23'}, {name: '15'}, {name: '10'}]


test('Snapshot test App component', () => {
  const snapshot = renderer
      .create(<App
          materials={data}
          sizes={sizes}
          matrix={[priceMatrixGold, priceMatrixPlatinum, priceMatrixSilver]}
      />)
      .toJSON()
  expect(snapshot).toMatchSnapshot()
})

test('Snapshot test Select component with materials', () => {
  const snapshot = renderer
      .create(<Select
          name={'Material'}
          options={data}
          selected = {0}
          onChange = {(e) => {}}
      />)
      .toJSON();
  expect(snapshot).toMatchSnapshot()
})

test('Snapshot test Select component with sizes', () => {
  const snapshot = renderer
      .create(<Select
          name={'Size'}
          options={sizes}
          selected = {0}
          onChange = {(e) => {}}
      />)
      .toJSON();
  expect(snapshot).toMatchSnapshot()
})

test('calculate price test1', () => {
  const colorIndex = 1
  const finenessIndex = 0
  const size = 23
  const engravingLength = 4
  expect(calculatePrice(colorIndex, finenessIndex, size, engravingLength, priceMatrixGold)).toBe(18440)
})

test('calculate price test2', () => {
  const colorIndex = 1
  const finenessIndex = 2
  const size = 0
  const engravingLength = 4
  expect(calculatePrice(colorIndex, finenessIndex, size, engravingLength, priceMatrixGold)).toBe(40)
})

//E2E TESTS


test('test material select', () => {
  const { getByTestId, getAllByTestId } = render(<App
      materials={data}
      sizes={sizes}
      matrix={[priceMatrixGold, priceMatrixPlatinum, priceMatrixSilver]}
  />)
  const select = getAllByTestId('select')[0]
  expect(select).toHaveValue('0')
  fireEvent.change(select, { target: { value: 2 } })
  expect(select).toHaveValue('2')
})

test('test color select', () => {
  const { getByTestId, getAllByTestId } = render(<App
      materials={data}
      sizes={sizes}
      matrix={[priceMatrixGold, priceMatrixPlatinum, priceMatrixSilver]}
  />)
  const select = getAllByTestId('select')[1]
  expect(select).toHaveValue('0')
  fireEvent.change(select, { target: { value: 1 } })
  expect(select).toHaveValue('1')
})

test('test fineness select', () => {
  const { getByTestId, getAllByTestId } = render(<App
      materials={data}
      sizes={sizes}
      matrix={[priceMatrixGold, priceMatrixPlatinum, priceMatrixSilver]}
  />)
  const select = getAllByTestId('select')[2]
  expect(select).toHaveValue('0')
  fireEvent.change(select, { target: { value: 2 } })
  expect(select).toHaveValue('2')
})

test('test sizes select', () => {
  const { getByTestId, getAllByTestId } = render(<App
      materials={data}
      sizes={sizes}
      matrix={[priceMatrixGold, priceMatrixPlatinum, priceMatrixSilver]}
  />)
  const select = getAllByTestId('select')[3]
  expect(select).toHaveValue('0')
  fireEvent.change(select, { target: { value: 1 } })
  expect(select).toHaveValue('1')
})

test('test reset selects after material changed', () => {
  const { getByTestId, getAllByTestId } = render(<App
      materials={data}
      sizes={sizes}
      matrix={[priceMatrixGold, priceMatrixPlatinum, priceMatrixSilver]}
  />)
  const selectM = getAllByTestId('select')[0]
  const selectC = getAllByTestId('select')[1]
  const selectF = getAllByTestId('select')[2]
  expect(selectM).toHaveValue('0')
  fireEvent.change(selectC, { target: { value: 1 } })
  fireEvent.change(selectF, { target: { value: 2 } })
  fireEvent.change(selectM, { target: { value: 1 } })
  expect(selectC).toHaveValue('0')
  expect(selectF).toHaveValue('0')
})

test('test input engraving', () => {
  const { getByTestId, getAllByTestId } = render(<App
      materials={data}
      sizes={sizes}
      matrix={[priceMatrixGold, priceMatrixPlatinum, priceMatrixSilver]}
  />)
  const input = getByTestId('input')

  expect(input).toHaveValue('')
  fireEvent.input(input, { target: { value: 'te st' } })

  expect(input).toHaveValue('te st')
})

