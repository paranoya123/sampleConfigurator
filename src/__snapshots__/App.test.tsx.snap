// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test App component 1`] = `
<div
  className="App"
>
  <label>
    Material
    <select
      data-testid="select"
      onChange={[Function]}
      value={0}
    >
      <option
        data-testid="select-option"
        value={0}
      >
         
        Gold
         
      </option>
      <option
        data-testid="select-option"
        value={1}
      >
         
        Platinum
         
      </option>
      <option
        data-testid="select-option"
        value={2}
      >
         
        Silver
         
      </option>
    </select>
  </label>
  <label>
    Color
    <select
      data-testid="select"
      onChange={[Function]}
      value={0}
    >
      <option
        data-testid="select-option"
        value={0}
      >
         
        Red
         
      </option>
      <option
        data-testid="select-option"
        value={1}
      >
         
        Green
         
      </option>
      <option
        data-testid="select-option"
        value={2}
      >
         
        Blue
         
      </option>
    </select>
  </label>
  <label>
    Fineness
    <select
      data-testid="select"
      onChange={[Function]}
      value={0}
    >
      <option
        data-testid="select-option"
        value={0}
      >
         
        500
         
      </option>
      <option
        data-testid="select-option"
        value={1}
      >
         
        585
         
      </option>
      <option
        data-testid="select-option"
        value={2}
      >
         
        700
         
      </option>
    </select>
  </label>
  <label>
    Sizes
    <select
      data-testid="select"
      onChange={[Function]}
      value={0}
    >
      <option
        data-testid="select-option"
        value={0}
      >
         
        23
         
      </option>
      <option
        data-testid="select-option"
        value={1}
      >
         
        15
         
      </option>
      <option
        data-testid="select-option"
        value={2}
      >
         
        10
         
      </option>
    </select>
  </label>
  <label>
    Engraving :
    <input
      data-testid="input"
      onChange={[Function]}
      type="text"
      value=""
    />
  </label>
  <label>
    Price: 
    0
  </label>
</div>
`;

exports[`Snapshot test Select component with materials 1`] = `
<label>
  Material
  <select
    data-testid="select"
    onChange={[Function]}
    value={0}
  >
    <option
      data-testid="select-option"
      value={0}
    >
       
      Gold
       
    </option>
    <option
      data-testid="select-option"
      value={1}
    >
       
      Platinum
       
    </option>
    <option
      data-testid="select-option"
      value={2}
    >
       
      Silver
       
    </option>
  </select>
</label>
`;

exports[`Snapshot test Select component with sizes 1`] = `
<label>
  Size
  <select
    data-testid="select"
    onChange={[Function]}
    value={0}
  >
    <option
      data-testid="select-option"
      value={0}
    >
       
      23
       
    </option>
    <option
      data-testid="select-option"
      value={1}
    >
       
      15
       
    </option>
    <option
      data-testid="select-option"
      value={2}
    >
       
      10
       
    </option>
  </select>
</label>
`;
